File: app/src/main/java/com/example/criminalintent/CrimePagerActivity.java
package com.example.criminalintent;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.viewpager.widget.ViewPager;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.button.MaterialButton;
import java.util.List;
import java.util.UUID;

public class CrimePagerActivity extends AppCompatActivity implements CrimeFragment.Callbacks {

    private static final String EXTRA_CRIME_ID =
            "com.bignerdranch.android.criminalintent.crime_id";

    private ViewPager mViewPager;
    private MaterialButton mJumpToFirstButton;
    private MaterialButton mJumpToLastButton;
    private List<Crime> mCrimes;

    public static Intent newIntent(Context packageContext, UUID crimeId) {
        Intent intent = new Intent(packageContext, CrimePagerActivity.class);
        intent.putExtra(EXTRA_CRIME_ID, crimeId);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_crime_pager);

        MaterialToolbar toolbar = findViewById(R.id.pager_toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        UUID crimeId = SerializationCompat.getSerializableExtra(getIntent(), EXTRA_CRIME_ID, UUID.class);

        mJumpToFirstButton = findViewById(R.id.jump_to_first);
        mJumpToLastButton = findViewById(R.id.jump_to_last);
        mViewPager = (ViewPager) findViewById(R.id.crime_view_pager);
        mCrimes = CrimeLab.get(this).getCrimes();
        FragmentManager fragmentManager = getSupportFragmentManager();

        mViewPager.setAdapter(new FragmentStatePagerAdapter(
                fragmentManager,
                FragmentStatePagerAdapter.BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
            @Override
            public Fragment getItem(int position) {
                Crime crime = mCrimes.get(position);
                return CrimeFragment.newInstance(crime.getId());
            }

            @Override
            public int getCount() {
                return mCrimes.size();
            }
        });

        mJumpToFirstButton.setOnClickListener(v -> mViewPager.setCurrentItem(0, false));
        mJumpToLastButton.setOnClickListener(v -> {
            if (!mCrimes.isEmpty()) {
                mViewPager.setCurrentItem(mCrimes.size() - 1, false);
            }
        });
        mViewPager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
            @Override
            public void onPageSelected(int position) {
                updateJumpButtons(position);
            }
        });

        if (crimeId != null) {
            for (int i = 0; i < mCrimes.size(); i++) {
                if (mCrimes.get(i).getId().equals(crimeId)) {
                    mViewPager.setCurrentItem(i, false);
                    break;
                }
            }
        }

        updateJumpButtons(mViewPager.getCurrentItem());
    }

    @Override
    public Intent getParentActivityIntent() {
        Intent intent = super.getParentActivityIntent();
        if (intent == null) {
            return null;
        }
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        return intent;
    }

    private void updateJumpButtons(int position) {
        int lastIndex = mCrimes.size() - 1;
        boolean hasCrimes = lastIndex >= 0;

        mJumpToFirstButton.setEnabled(hasCrimes && position > 0);
        mJumpToLastButton.setEnabled(hasCrimes && position < lastIndex);
    }

    @Override
    public void onCrimeUpdated(Crime crime) {
    }

    @Override
    public void onCrimeDeleted(Crime crime) {
        finish();
    }
}

File: app/src/main/java/com/bignerdranch/android/criminalintent/CrimePagerActivity.kt
package com.bignerdranch.android.criminalintent

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.ViewPager
import java.util.UUID

class CrimePagerActivity : AppCompatActivity(), CrimeFragment.Callbacks {

    private lateinit var viewPager: ViewPager
    private lateinit var crimes: List<Crime>
    private lateinit var jumpToFirstButton: Button
    private lateinit var jumpToLastButton: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_crime_pager)

        val crimeId = intent.getSerializableExtra(EXTRA_CRIME_ID) as UUID

        viewPager = findViewById(R.id.crime_view_pager)
        jumpToFirstButton = findViewById(R.id.jump_to_first_button)
        jumpToLastButton = findViewById(R.id.jump_to_last_button)

        crimes = CrimeLab.get(this).getCrimes()

        val fragmentManager = supportFragmentManager
        viewPager.adapter = object : FragmentStatePagerAdapter(fragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
            override fun getItem(position: Int): Fragment {
                val crime = crimes[position]
                return CrimeFragment.newInstance(crime.id)
            }

            override fun getCount(): Int {
                return crimes.size
            }
        }

        viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

            override fun onPageSelected(position: Int) {
                updateButtons(position)
            }

            override fun onPageScrollStateChanged(state: Int) {}
        })

        jumpToFirstButton.setOnClickListener {
            viewPager.currentItem = 0
        }

        jumpToLastButton.setOnClickListener {
            viewPager.currentItem = crimes.size - 1
        }

        for (i in crimes.indices) {
            if (crimes[i].id == crimeId) {
                viewPager.currentItem = i
                updateButtons(i)
                break
            }
        }
    }

    override fun onCrimeUpdated(crime: Crime) {
        // Optional: Implement if pager needs to react to crime updates
    }

    private fun updateButtons(position: Int) {
        jumpToFirstButton.isEnabled = position > 0
        jumpToLastButton.isEnabled = position < crimes.size - 1
    }

    companion object {
        private const val EXTRA_CRIME_ID = "com.bignerdranch.android.criminalintent.crime_id"

        fun newIntent(packageContext: Context, crimeId: UUID): Intent {
            val intent = Intent(packageContext, CrimePagerActivity::class.java)
            intent.putExtra(EXTRA_CRIME_ID, crimeId)
            return intent
        }
    }
}
